File: platform/polyglot/langgraph-flow-serializer/py/langgraph_flow_serializer/adapters/flowise_adapter.py
"""
Flowise dialect adapter.

Mirrors mjs/src/adapters/flowise.adapter.ts.

Flowise uses:
  - data.name: camelCase LangChain.js class name
  - data.inputs: flat dict of input values
  - No data.node.template structure

Detection heuristic: first node has data.name as camelCase string
and does NOT have a data.node.template structure.
"""

from __future__ import annotations

import re
from typing import Any

from langgraph_flow_serializer.adapters.base import DialectAdapter
from langgraph_flow_serializer.types import AIEdge, AINode, AINodeData, AIWorkflow


def _is_camel_case(s: str) -> bool:
    return bool(re.match(r"^[a-z][a-zA-Z0-9]*$", s))


class FlowiseAdapter(DialectAdapter):
    def detect(self, json: Any) -> bool:
        if not isinstance(json, dict):
            return False
        nodes = json.get("nodes")
        if not isinstance(nodes, list) or len(nodes) == 0:
            return False
        first = nodes[0]
        if not isinstance(first, dict):
            return False
        data = first.get("data")
        if not isinstance(data, dict):
            return False
        # Langflow has data.node.template — reject those
        if isinstance(data.get("node"), dict) and isinstance(
            data["node"].get("template"), dict
        ):
            return False
        name = data.get("name")
        return isinstance(name, str) and _is_camel_case(name)

    def from_dialect(self, dialect_json: Any) -> AIWorkflow:
        nodes: list[AINode] = []
        for fn in dialect_json.get("nodes", []):
            data = fn.get("data") or {}
            node_data = AINodeData(
                node_type=data.get("name", ""),
                category=data.get("category", "Uncategorized"),
                name=data.get("label") or data.get("name"),
                inputs=data.get("inputs") or {},
                credentials=data.get("credentials") or None,
            )
            nodes.append(
                AINode(
                    id=fn["id"],
                    type=fn.get("type", "customNode"),
                    position=fn.get("position") or {"x": 0, "y": 0},
                    data=node_data,
                )
            )

        edges: list[AIEdge] = []
        for fe in dialect_json.get("edges", []):
            edges.append(
                AIEdge(
                    id=fe["id"],
                    source=fe["source"],
                    source_handle=fe.get("sourceHandle") or f"{fe['source']}-output",
                    target=fe["target"],
                    target_handle=fe.get("targetHandle") or f"{fe['target']}-input",
                    type=fe.get("type") or None,
                )
            )

        vp_raw = dialect_json.get("viewport")
        from langgraph_flow_serializer.types import AIViewport

        viewport = (
            AIViewport(x=vp_raw["x"], y=vp_raw["y"], zoom=vp_raw["zoom"])
            if isinstance(vp_raw, dict)
            else None
        )

        return AIWorkflow(
            id=dialect_json.get("id") or None,
            name=dialect_json.get("name") or "Untitled Flowise Flow",
            description=dialect_json.get("description") or None,
            viewport=viewport,
            nodes=nodes,
            edges=edges,
        )

    def to_dialect(self, workflow: AIWorkflow) -> dict:
        nodes = []
        for n in workflow.nodes:
            node: dict = {
                "id": n.id,
                "type": n.type,
                "position": n.position,
                "data": {
                    "name": n.data.node_type,
                    "label": n.data.name or n.data.node_type,
                    "category": n.data.category,
                    "inputs": n.data.inputs,
                },
            }
            if n.data.credentials:
                node["data"]["credentials"] = n.data.credentials
            nodes.append(node)

        edges = [
            {
                "id": e.id,
                "source": e.source,
                "sourceHandle": e.source_handle,
                "target": e.target,
                "targetHandle": e.target_handle,
                **({"type": e.type} if e.type else {}),
            }
            for e in workflow.edges
        ]

        result: dict = {"name": workflow.name, "nodes": nodes, "edges": edges}
        if workflow.id:
            result["id"] = workflow.id
        if workflow.description:
            result["description"] = workflow.description
        if workflow.viewport:
            result["viewport"] = {
                "x": workflow.viewport.x,
                "y": workflow.viewport.y,
                "zoom": workflow.viewport.zoom,
            }
        return result

File: platform/polyglot/langgraph-flow-serializer/mjs/src/adapters/flowise.adapter.ts
/**
 * Flowise dialect adapter.
 *
 * Flowise uses `data.name` (camelCase LangChain.js class name) and a flat
 * `data.inputs` object. There is no `data.node.template` structure.
 *
 * Detection heuristic: nodes have `data.name` as a camelCase string
 * and do NOT have `data.node.template`.
 */

import type { AIWorkflow, AINode, AIEdge } from "../types.js";
import type { DialectAdapter } from "./adapter.js";

interface FlowiseNodeData {
  name?: string;
  label?: string;
  category?: string;
  inputs?: Record<string, unknown>;
  credentials?: Record<string, string>;
  [key: string]: unknown;
}

interface FlowiseNode {
  id: string;
  type?: string;
  position?: { x: number; y: number };
  data?: FlowiseNodeData;
  [key: string]: unknown;
}

interface FlowiseEdge {
  id: string;
  source: string;
  sourceHandle?: string;
  target: string;
  targetHandle?: string;
  type?: string;
  [key: string]: unknown;
}

interface FlowiseFlow {
  id?: string;
  name?: string;
  description?: string;
  viewport?: { x: number; y: number; zoom: number };
  nodes?: FlowiseNode[];
  edges?: FlowiseEdge[];
  [key: string]: unknown;
}

function isCamelCase(str: string): boolean {
  // camelCase: starts with lowercase, contains no spaces
  return /^[a-z][a-zA-Z0-9]*$/.test(str);
}

export const flowiseAdapter: DialectAdapter = {
  detect(json: unknown): boolean {
    if (typeof json !== "object" || json === null) return false;
    const flow = json as FlowiseFlow;
    const nodes = flow.nodes;
    if (!Array.isArray(nodes) || nodes.length === 0) return false;

    // Check first node — Flowise uses data.name (camelCase) without data.node.template
    const firstNode = nodes[0];
    if (typeof firstNode?.data !== "object" || firstNode.data === null) return false;
    const data = firstNode.data as FlowiseNodeData;
    if ("node" in data && typeof (data as Record<string, unknown>).node === "object") {
      return false; // Langflow signature
    }
    return typeof data.name === "string" && isCamelCase(data.name);
  },

  fromDialect(dialectJson: unknown): AIWorkflow {
    const flow = dialectJson as FlowiseFlow;
    const nodes: AINode[] = (flow.nodes ?? []).map((fn): AINode => {
      const data = fn.data ?? {};
      return {
        id: fn.id,
        type: fn.type ?? "customNode",
        position: fn.position ?? { x: 0, y: 0 },
        data: {
          nodeType: data.name ?? "",
          category: data.category ?? "Uncategorized",
          name: data.label ?? data.name,
          inputs: data.inputs ?? {},
          ...(data.credentials ? { credentials: data.credentials } : {}),
        },
      };
    });

    const edges: AIEdge[] = (flow.edges ?? []).map((fe): AIEdge => ({
      id: fe.id,
      source: fe.source,
      sourceHandle: fe.sourceHandle ?? `${fe.source}-output`,
      target: fe.target,
      targetHandle: fe.targetHandle ?? `${fe.target}-input`,
      ...(fe.type ? { type: fe.type } : {}),
    }));

    return {
      ...(flow.id ? { id: flow.id } : {}),
      name: flow.name ?? "Untitled Flowise Flow",
      ...(flow.description ? { description: flow.description } : {}),
      ...(flow.viewport ? { viewport: flow.viewport } : {}),
      nodes,
      edges,
    };
  },

  toDialect(workflow: AIWorkflow): FlowiseFlow {
    const nodes: FlowiseNode[] = workflow.nodes.map((n): FlowiseNode => ({
      id: n.id,
      type: n.type,
      position: n.position,
      data: {
        name: n.data.nodeType,
        label: n.data.name ?? n.data.nodeType,
        category: n.data.category,
        inputs: n.data.inputs,
        ...(n.data.credentials ? { credentials: n.data.credentials } : {}),
      },
    }));

    const edges: FlowiseEdge[] = workflow.edges.map((e): FlowiseEdge => ({
      id: e.id,
      source: e.source,
      sourceHandle: e.sourceHandle,
      target: e.target,
      targetHandle: e.targetHandle,
      ...(e.type ? { type: e.type } : {}),
    }));

    return {
      ...(workflow.id ? { id: workflow.id } : {}),
      name: workflow.name,
      ...(workflow.description ? { description: workflow.description } : {}),
      ...(workflow.viewport ? { viewport: workflow.viewport } : {}),
      nodes,
      edges,
    };
  },
};
